runtimes/huggingface/mlserver_huggingface/codecs.py

```diff
@@ -1,89 +1,351 @@
-from typing import Optional, Type, Any, Dict
+import json
+from typing import Optional, Type, Any, Dict, List, Union
+from functools import partial
+from xmlrpc.client import Boolean
 from mlserver.codecs.utils import (
+    InputOrOutput,
     has_decoded,
     _save_decoded,
     get_decoded_or_raw,
+    is_single_value,
+    set_content_type,
 )
+
+from mlserver.codecs.lists import is_list_of
 from mlserver.codecs.base import (
     RequestCodec,
     register_request_codec,
+    register_input_codec,
+    find_input_codec,
+    find_input_codec_by_payload,
     InputCodec as InputCodecTy,
 )
-from mlserver.codecs import StringCodec
+from mlserver.codecs import (
+    InputCodec,
+    StringCodec,
+)
+from mlserver.codecs.base64 import _decode_base64, _encode_base64
+from mlserver.codecs.lists import as_list
+from mlserver.types import RequestInput, ResponseOutput
+from PIL import Image
+
 from mlserver.types import (
     InferenceRequest,
     InferenceResponse,
 )
+from transformers.pipelines import Conversation
+
+from mlserver.types.dataplane import Parameters
+from .common import (
+    CommonJSONEncoder,
+    CONTENT_TYPE_JSON,
+    CONTENT_TYPE_IMAGE,
+    CONTENT_TYPE_CONVERSATION,
+    deserialize_conversation,
+    get_img_bytes,
+    build_image,
+    metainfo_for,
+)
 
 
-class MultiInputRequestCodec(RequestCodec):
+def get_input_codec(item: InputOrOutput, default: Optional[Type[InputCodecTy]]):
+    if item.parameters is None:
+        return default
+    if not item.parameters.content_type:
+        return default
+    return find_input_codec(item.parameters.content_type)
+
+
+def find_input_codec_by_payload_content(data: Any):
     """
-    The SingleInputRequestCodec can be used as a "meta-implementation" for other
-    codecs. Its goal to decode the whole request simply as the first decoded
-    element.
+    for huggingface JSONCodec has high priority
     """
+    if JSONCodec.can_encode(data):
+        return JSONCodec
+    if ImageCodec.can_encode(data):
+        return ImageCodec
+    if ConversationCodec.can_encode(data):
+        return ConversationCodec
+    return find_input_codec_by_payload(data)
+
+
+@register_input_codec
+class ImageCodec(InputCodecTy):
```

Comment:
I was thinking that, to help future maintainers, we should split these codecs into their own files. As in, instead of keeping them all under `mlserver_huggingface/codecs.py`, we should create a new `mlserver_huggingface/codecs` package (i.e. folder), and then have:

- `mlserver_huggingface/codecs/image.py`
- `mlserver_huggingface/codecs/conversation.py`
- `mlserver_huggingface/codecs/json.py`
- ....

Each of these files could also have any codec-specific helpers (like `get_img_bytes`, which could live in `mlserver_huggingface/codecs/image.py`). 

This will also help with the code review :+1:

Reply:
Your suggestion is right. I will revise it according to that